python/lsst/ctrl/bps/wms/panda/edgenode/cmd_line_decoder.py

```diff
@@ -4,12 +4,30 @@
 cluster management -> Edge node -> Container. Once the command line hexified it can contain symbols often
 stripped such as ',",` and multiline commands.
 """
+import os
+import re
 import sys
 import binascii
-cmdline = str(binascii.unhexlify(sys.argv[1]).decode())
-dataparams = sys.argv[2].split("+")
-cmdline = cmdline.replace("${filename}", dataparams[0])
-if len(dataparams) > 2:
-    cmdline = cmdline.replace("${qgraph-id}", dataparams[2])
-    cmdline = cmdline.replace("${qgraph-node-id}", dataparams[3])
-print(cmdline)
+
+
+def replace_environment_vars(cmdline):
+    envs_to_replace = re.findall(r'<ENV:(.*?)>', cmdline)
+    for env_var in envs_to_replace:
+        if os.getenv(env_var):
+            cmdline = cmdline.replace('<ENV:'+env_var+'>', os.getenv(env_var))
```

Comment:
I'm confused because it looks like this env var replacement code is not using the envvar replacement code from earlier. Surely there should only be one place in BPS that knows how to replace the special BPS `<ENV` syntax with the environment variable value.

Reply:
This part in to be executed on the computing node, another parts to be executed on the submission side, they are running in different locations.

Reply:
Time of execution does not explain logic duplication though. It seems to me that env-replacement is a generic concept that deserves a single implementation location.

Reply:
Replacing the <ENV syntax for inside jobs is plugin specific (e.g., HTCondor wants parens instead of braces).  Otherwise we should just have bps put normal shell syntax in the GenericWorkflow command lines.

Reply:
We can't initialize the BPS code in the light weight execution wrapper, this is why this code is separated. Another thing is that submission side replaces the lazy parameters which are static or should be initialized on the submission node, and the edge node performs the actual lazy variables resolution.

Reply:
The docstring at the top should be very clear about why this file doesn't depend on any code from BPS.

My comment about `<ENV` syntax is that this is implementing the generic solution of replacing the `ENV` string with the value of the environment variable.